File: frontend/src/types/todo.ts
/**
 * TODO 类型定义
 * 
 * 对照后端契约: backend/core/contracts/todo_models.py
 * 对照文档: .kiro/docs/api_contracts.md §3.3
 */

export type TodoType = "todo" | "done";
export type TodoStatus = "pending" | "completed" | "cancelled";

// ==================== 常量 ====================

export const TODO_CONTENT_MAX_LENGTH = 100;
export const TODO_DAILY_LIMIT = 50;

// ==================== 实体类型 ====================

export interface TodoEntry {
  todo_id: string;
  user_id: string;
  content: string;
  entry_type: TodoType;
  status: TodoStatus;
  created_at: string;
  completed_at?: string;
  extracted_intent?: Record<string, unknown>;
  linked_insight_id?: string;
}

// ==================== 请求类型 ====================

export interface TodoCreateRequest {
  content: string;
  entry_type: TodoType;
}

export interface TodoUpdateRequest {
  content?: string;
  status?: TodoStatus;
}

export interface TodoListParams {
  date?: string;           // YYYY-MM-DD，默认今天
  entry_type?: TodoType;
  status?: TodoStatus;
  limit?: number;          // 默认 50
  offset?: number;
}

// ==================== 响应类型 ====================

export interface TodoListResponse {
  entries: TodoEntry[];
  total: number;
  has_more: boolean;
}

export interface TodoDailySummary {
  user_id: string;
  date: string;
  total_count: number;
  todo_count: number;
  done_count: number;
  completed_count: number;
  entries: TodoEntry[];
}

// ==================== 辅助函数 ====================

/**
 * 校验 TODO 内容长度
 */
export function validateTodoContent(content: string): boolean {
  const trimmed = content.trim();
  return trimmed.length >= 1 && trimmed.length <= TODO_CONTENT_MAX_LENGTH;
}

/**
 * 获取内容长度信息
 */
export function getContentLengthInfo(content: string): {
  current: number;
  max: number;
  remaining: number;
  isValid: boolean;
} {
  const current = content.trim().length;
  return {
    current,
    max: TODO_CONTENT_MAX_LENGTH,
    remaining: TODO_CONTENT_MAX_LENGTH - current,
    isValid: current >= 1 && current <= TODO_CONTENT_MAX_LENGTH,
  };
}

File: backend/models/todo.py
"""
TODO Database Models

TODO/DONE 行为记录的 SQLAlchemy 模型。

对照契约: backend/core/contracts/todo_models.py
"""

from datetime import datetime
from typing import Optional

from sqlalchemy import Column, DateTime, Enum, Index, JSON, String, Text
from sqlalchemy.orm import relationship

from backend.db.base import Base
from backend.core.contracts.todo_models import TodoType, TodoStatus


class TodoEntry(Base):
    """
    TODO 条目表
    
    存储用户的每日行为记录（TO-DO/DONE）。
    每条记录限制在 0-100 字以内。
    """
    __tablename__ = "todo_entries"

    # 主键
    todo_id = Column(
        String(64), 
        primary_key=True,
        comment="TODO ID，格式 todo_xxxxxxxxxxxx"
    )
    
    # 关联用户
    user_id = Column(
        String(64), 
        nullable=False, 
        index=True,
        comment="用户ID"
    )
    
    # 内容（限制100字）
    content = Column(
        String(100),
        nullable=False,
        comment="内容，最大100字符"
    )
    
    # 条目类型
    entry_type = Column(
        Enum(TodoType),
        nullable=False,
        comment="类型: todo/done"
    )
    
    # 状态
    status = Column(
        Enum(TodoStatus),
        nullable=False,
        default=TodoStatus.PENDING,
        comment="状态: pending/completed/cancelled"
    )
    
    # 时间戳
    created_at = Column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
        index=True,
        comment="创建时间"
    )
    
    completed_at = Column(
        DateTime,
        nullable=True,
        comment="完成时间"
    )
    
    # 批处理填充字段
    extracted_intent = Column(
        JSON,
        nullable=True,
        comment="LLM 提取的意图结构"
    )
    
    linked_insight_id = Column(
        String(64),
        nullable=True,
        comment="关联的 Insight ID"
    )

    # 复合索引：用户ID + 创建日期（用于日期查询）
    __table_args__ = (
        Index('ix_todo_user_date', 'user_id', 'created_at'),
    )

    def __repr__(self) -> str:
        return f"<TodoEntry {self.todo_id} type={self.entry_type}>"

    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "todo_id": self.todo_id,
            "user_id": self.user_id,
            "content": self.content,
            "entry_type": self.entry_type.value if self.entry_type else None,
            "status": self.status.value if self.status else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "extracted_intent": self.extracted_intent,
            "linked_insight_id": self.linked_insight_id,
        }
